crates/vm/src/expr.rs

```diff
@@ -1611,12 +1741,14 @@ impl QueryExpr {
                     q = Self::optimize_select(q, op, &tables);
                 }
                 Query::JoinInner(join) => {
-                    q = q.with_join_inner(join.rhs.optimize(row_count), join.col_lhs, join.col_rhs)
+                    q = q.with_join_inner(join.rhs.optimize(row_count), join.col_lhs, join.col_rhs, join.semi);
                 }
                 _ => q.query.push(query),
             };
         }
 
+        // Make sure to `try_semi_join` before `try_index_join`, as the latter depends on the former.
+        let q = q.try_semi_join();
```

Comment:
Ok I think I understand now. The compilation path is now
```
sql -> inner join + project -> semi join -> index join
```
And then for eval_incr, `to_inner_join` will now take an index join and produce an `inner_join{semi: true}`, whereas before it would produce an `inner_join + project`

Is this correct?

Reply:
Yes, that's correct.

Reply:
I expanded the comment to that effect.